File: LeetCodeSolutions/LeetCode_0215.py
from typing import List
import random


class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        nums.sort()
        return nums[-k]

    def findKthLargest_partition(self, nums: List[int], k: int) -> int:
        """
        partition
        Top-K algorithm
        """

        k = len(nums) - k
        if k < 0:
            return None

        def __partition(low, high):
            random_idx = random.randint(low, high)
            nums[low], nums[random_idx] = nums[random_idx], nums[low]
            x = nums[low]
            i = low
            j = high
            while i < j:
                while i <= j and nums[i] <= x:
                    i += 1
                while j >= i and nums[j] >= x:
                    j -= 1
                if i >= j:
                    break
                nums[i], nums[j] = nums[j], nums[i]
            nums[j], nums[low] = nums[low], nums[j]
            return j

        low, high = 0, len(nums) - 1
        while low < high:
            pivot = __partition(low, high)
            if pivot == k:
                break
            elif pivot < k:
                low = pivot + 1
            else:
                high = pivot - 1
        return nums[k]

File: LeetCodeSolutions/LeetCode_0215.cpp
class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {
        return quickSelect(nums, 0, nums.size() - 1, nums.size() - k);
    }
    void print(vector<int> & nums){
        for(auto n: nums)
            cout << n << ' ';
        cout << endl;
    }

    int quickSelect(vector<int> &a, int l, int r, int k){
        randomPartition(a, l, r);
        int p = Partition(a, l, r);
        // print(a);
        if(p == k) return a[p];
        return p < k? quickSelect(a, p + 1, r, k) : quickSelect(a, l, p - 1, k);
    }

    void randomPartition(vector<int> &a, int l, int r){
        int k = rand() % (r - l + 1) + l;
        swap(a[r], a[k]);
    }

    int Partition(vector<int>& a, int l, int r) {
        int x = a[r], i = l - 1;
        for (int j = l; j < r; ++j) {
            if (a[j] <= x) {
                swap(a[++i], a[j]);
            }
        }
        swap(a[i + 1], a[r]);
        return i + 1;
    }
};